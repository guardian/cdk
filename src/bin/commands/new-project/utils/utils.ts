import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
