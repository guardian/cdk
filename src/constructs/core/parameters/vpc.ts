import { GuParameter } from "./base";
import type { GuStack } from "../stack";
import type { GuNoTypeParameterProps } from "./base";

export class GuSubnetListParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "List<AWS::EC2::Subnet::Id>" });
  }
}

export class GuVpcParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::EC2::VPC::Id",
    });
  }
}
