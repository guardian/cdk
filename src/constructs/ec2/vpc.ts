import { Subnet, Vpc } from "aws-cdk-lib/aws-ec2";
import type { ISubnet, IVpc, VpcAttributes } from "aws-cdk-lib/aws-ec2";
import { GuSubnetListParameter, GuVpcParameter } from "../core";
import type { GuStack } from "../core";

interface VpcFromIdProps extends Omit<VpcAttributes, "availabilityZones"> {
  availabilityZones?: string[];
}

type VpcFromIdParameterProps = Omit<VpcFromIdProps, "vpcId">;

// TODO convert to boolean
export enum SubnetType {
  PUBLIC = "Public",
  PRIVATE = "Private",
}

// TODO: Migrate this to use `AppIdentity`
export interface GuSubnetProps {
  type?: SubnetType;
  app?: string;
}

/*
 * Some stacks maybe have multiple patterns (such as `GuEc2App`), therefore creating multiple VPC and subnet resources.
 * This function ensures that, where desired, we can have unique IDs for each VPC and subnet declaration
 */
export const maybeApp = (props?: { app?: string }): string => props?.app ?? "";

export class GuVpc {
  static subnets(scope: GuStack, subnets: string[]): ISubnet[] {
    /**
     *  As of ~1.77.0 of the aws/cdk, this line
     * [ previously Subnet.fromSubnetId(scope, `subnet-${subnetId}`, subnetId)) ]
     * resulted in the following error
     * "Found an encoded list token string in a scalar string context. Use 'Fn.select(0, list)' (not 'list[0]') to extract elements from token lists."
     * See https://github.com/aws/aws-cdk/issues/11945 for more information
     *
     * As a hacky workaround, we have moved to using the fromSubnetAttributes method and hardcoded an empty value
     * for the routeTableId prop. This prevents the error and, when tested on existing stacks, results in no change to the CFN output
     *
     * TODO: Understand VPCs and Subnets better and develop a better solution to this problem
     */
    return subnets.map((subnetId) =>
      Subnet.fromSubnetAttributes(scope, `subnet-${subnetId}`, { subnetId, routeTableId: " " })
    );
  }

  static subnetsFromParameter(scope: GuStack, props?: GuSubnetProps): ISubnet[] {
    const type = props?.type ?? SubnetType.PRIVATE;

    const subnets = new GuSubnetListParameter(scope, `${maybeApp(props)}${type}Subnets`, {
      description: `A list of ${type.toLowerCase()} subnets`,

      // TODO use `SSM_PARAMETER_PATHS.PrimaryVpcPrivateSubnets` or `SSM_PARAMETER_PATHS.PrimaryVpcPublicSubnets`
      default: `/account/vpc/primary/subnets/${type.toLowerCase()}`,
      fromSSM: true,
    });

    return GuVpc.subnets(scope, subnets.valueAsList);
  }

  static fromId(scope: GuStack, id: string, props: VpcFromIdProps): IVpc {
    return Vpc.fromVpcAttributes(scope, id, {
      availabilityZones: scope.availabilityZones,
      ...props,
    });
  }

  static fromIdParameter(scope: GuStack, id: string, props?: VpcFromIdParameterProps): IVpc {
    const vpc = GuVpcParameter.getInstance(scope);
    return GuVpc.fromId(scope, id, { ...props, vpcId: vpc.valueAsString });
  }
}
