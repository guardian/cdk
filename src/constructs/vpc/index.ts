export * from "./vpc";
export * from "./vpc-import";
