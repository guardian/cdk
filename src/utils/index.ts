export * as array from "./array";
export * as exec from "./exec";
export * as git from "./git";
export * as math from "./math";
export * as singleton from "./singleton";
