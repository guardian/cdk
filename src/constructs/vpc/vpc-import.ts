import { Aws, Fn, Token } from "aws-cdk-lib";
import type { ISubnet, IVpc } from "aws-cdk-lib/aws-ec2";
import { Vpc } from "aws-cdk-lib/aws-ec2";
import { isSingletonPresentInStack } from "../../utils/singleton";
import type { GuStack } from "../core";
import { GuVpcParameter, GuVpcPrivateSubnetsParameter, GuVpcPublicSubnetsParameter } from "../core";

/**
 * A singleton class that imports a VPC, including the public and private subnets, from SSM Parameters.
 *
 * By default, the following SSM Parameters are used:
 * - [`/account/vpc/primary/id` for the VPC ID]{@link GuVpcParameter}
 * - [`/account/vpc/primary/subnets/public` for the public subnets]{@link GuVpcPublicSubnetsParameter}
 * - [`/account/vpc/primary/subnets/private` for the private subnets]{@link GuVpcPrivateSubnetsParameter}
 *
 * Account VPCs and related SSM Parameters are managed in https://github.com/guardian/aws-account-setup.
 */
export class GuVpcImport {
  // This property wouldn't really be used, but its necessary for the singleton implementation.
  public readonly stack: GuStack;

  /**
   * The imported VPC.
   */
  public readonly vpc: IVpc;

  /**
   * The imported public subnets.
   * This property can also be read from the [vpc]{@link GuVpcImport#vpc} property.
   */
  public readonly publicSubnets: ISubnet[];

  /**
   * The imported private subnets.
   * This property can also be read from the [vpc]{@link GuVpcImport#vpc} property.
   */
  public readonly privateSubnets: ISubnet[];

  private static instance: GuVpcImport | undefined;

  private constructor(scope: GuStack, vpc: IVpc) {
    const { publicSubnets, privateSubnets } = vpc;

    this.stack = scope;

    this.vpc = vpc;
    this.publicSubnets = publicSubnets;
    this.privateSubnets = privateSubnets;
  }

  /**
   * Imports a VPC (including the public and private subnets) from SSM Parameters.
   */
  static fromSsmParameters(stack: GuStack): GuVpcImport {
    if (!this.instance || !isSingletonPresentInStack(stack, this.instance)) {
      const vpcId = GuVpcParameter.getInstance(stack).valueAsString;
      const publicSubnetIds = GuVpcPublicSubnetsParameter.getInstance(stack).valueAsList;
      const privateSubnetIds = GuVpcPrivateSubnetsParameter.getInstance(stack).valueAsList;

      const vpc = Vpc.fromVpcAttributes(stack, "VpcImport", {
        /*
        AWS CDK requires this property to be set.
        We're pulling the subnets from SSM Parameters at runtime into s list.
        At compile time, we can only assert the list has 1 item (though in reality, we know it has 3 items).
        AWS CDK wants `publicSubnetIds` and `privateSubnetIds` to be a multiple of `availabilityZones`.

        If we were to use the dynamic value of `scope.availabilityZones` we get the error:
          > Error: Number of privateSubnetIds (1) must be a multiple of availability zones (2).

        Set the number of AZs to 1 to avoid the error.
        */
        availabilityZones: [Aws.NO_VALUE],

        vpcId,
        publicSubnetIds,
        privateSubnetIds,
      });

      this.instance = new GuVpcImport(stack, vpc);
    }

    return this.instance;
  }

  /**
   * Imports a VPC from the `eu-west-1` region (including the public and private subnets) from SSM Parameters.
   * Use this if you need to know the availability zones of the VPC at compile time.
   *
   * IMPORTANT: The order of the subnets in the SSM Parameters must match the order of the availability zones.
   * Prefer [fromSsmParameters]{@link GuVpcImport#fromSsmParameters}.
   */
  static fromSsmParametersRegional(stack: GuStack): GuVpcImport {
    if (!this.instance || !isSingletonPresentInStack(stack, this.instance)) {
      const { region } = stack;

      if (Token.isUnresolved(region) || region !== "eu-west-1") {
        throw new Error("This function is only supported in the eu-west-1 region");
      }

      const availabilityZones = ["eu-west-1a", "eu-west-1b", "eu-west-1c"];

      const vpcId = GuVpcParameter.getInstance(stack).valueAsString;
      const publicSubnetIds = GuVpcPublicSubnetsParameter.getInstance(stack).valueAsList;
      const privateSubnetIds = GuVpcPrivateSubnetsParameter.getInstance(stack).valueAsList;

      const vpc = Vpc.fromVpcAttributes(stack, "VpcImport", {
        availabilityZones,
        vpcId,
        publicSubnetIds: availabilityZones.map((_, index) => Fn.select(index, publicSubnetIds)),
        privateSubnetIds: availabilityZones.map((_, index) => Fn.select(index, privateSubnetIds)),
      });

      this.instance = new GuVpcImport(stack, vpc);
    }

    return this.instance;
  }
}
