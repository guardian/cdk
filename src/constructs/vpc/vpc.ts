import type { VpcProps } from "@aws-cdk/aws-ec2";
import { GatewayVpcEndpointAwsService, SubnetType, Vpc } from "@aws-cdk/aws-ec2";
import { StringListParameter, StringParameter } from "@aws-cdk/aws-ssm";
import { VPC_SSM_PARAMETER_PREFIX } from "../../constants";
import { GuStatefulMigratableConstruct } from "../../utils/mixin";
import type { GuMigratingResource, GuStack } from "../core";

export interface GuVpcCustomProps {
  /**
   * Whether to add SSM Parameters containing VPC metadata, which are expected
   * to exist by many other Guardian CDK patterns.
   *
   * Defaults to 'true'.
   */
  ssmParameters?: boolean;

  /**
   * An identifier for the VPC to namespace SSM parameters. Customise when you
   * have multiple teams/VPCs in the same account.
   *
   * This will be combined with the /account/vpc prefix for the full parameter
   * name. e.g. '/account/vpc/primary'.
   *
   * Defaults to 'primary'.
   */
  ssmParametersNamespace?: string;
}

export interface GuVpcProps extends GuVpcCustomProps, VpcProps, GuMigratingResource {}

/**
 * Construct which creates a Virtual Private Cloud.
 *
 * NOTE: this construct requires an
 * [environment](https://docs.aws.amazon.com/cdk/latest/guide/environments.html)
 * to be set to function correctly. Without this, an environment-agnostic
 * template will be produced, which will only use two AZs even if the region
 * contains more than that.
 *
 * The VPC is provisioned with a public and private subnet for each availability
 * zone, with IPs spread evenly across these. Instances in private subnets
 * cannot be accessed directly over the internet. You should locate services in
 * instances in the private subnets, and provide HTTP(S) as required via load
 * balancers living in the public subnets.
 *
 * A CIDR block is a combined IP and network mask that specifies the IP
 * addresses in your VPC. For testing purposes any (AWS supported) CIDR block
 * will do, but for production you should:
 *
 * - use a /21 network mask (providing up to 2048 IPs, which is plenty for most
 *   accounts)
 * - ask Enterprise Tech for an IP address to go with this range (so that
 *   peering with other company VPCs is possible without IP clashes)
 *
 * A managed NAT is created for each private subnet to allow access to the
 * internet.
 *
 * For more information on VPCs and AWS see:
 * https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Subnets.html.
 */
export class GuVpc extends GuStatefulMigratableConstruct(Vpc) {
  constructor(scope: GuStack, id: string, props?: GuVpcProps) {
    const defaultVpcProps: VpcProps = {
      gatewayEndpoints: {
        s3: {
          service: GatewayVpcEndpointAwsService.S3,
        },
        dynamodb: {
          service: GatewayVpcEndpointAwsService.DYNAMODB,
        },
      },

      subnetConfiguration: [
        {
          name: "ingress",
          subnetType: SubnetType.PUBLIC,
        },
        {
          name: "application",
          subnetType: SubnetType.PRIVATE_WITH_NAT,
        },
      ],
    };

    const defaultCustomProps: Required<GuVpcCustomProps> = {
      ssmParameters: true,
      ssmParametersNamespace: "primary",
    };

    super(scope, id, { ...defaultVpcProps, ...props });

    if (props?.ssmParameters ?? defaultCustomProps.ssmParameters) {
      const namespace = props?.ssmParametersNamespace ?? defaultCustomProps.ssmParametersNamespace;
      const prefix = `${VPC_SSM_PARAMETER_PREFIX}/${namespace}`;

      new StringParameter(scope, "vpcID", {
        parameterName: `${prefix}/id`,
        stringValue: this.vpcId,
      });

      new StringListParameter(scope, "publicSubnets", {
        parameterName: `${prefix}/subnets/public`,
        stringListValue: this.publicSubnets.map((subnet) => subnet.subnetId),
      });

      new StringListParameter(scope, "privateSubnets", {
        parameterName: `${prefix}/subnets/private`,
        stringListValue: this.privateSubnets.map((subnet) => subnet.subnetId),
      });
    }
  }
}
