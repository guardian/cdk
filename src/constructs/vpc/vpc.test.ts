import "@aws-cdk/assert/jest";
import "../../utils/test/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { simpleGuStackForTesting } from "../../utils/test";
import { GuVpc } from "./vpc";

describe("The GuVpc construct", () => {
  it("should match snapshot", () => {
    const stack = simpleGuStackForTesting({ stack: "test-stack" });
    new GuVpc(stack, "MyVpc");
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });

  it("should create VPC SSM parameters by default", () => {
    const stack = simpleGuStackForTesting({ stack: "test-stack" });
    new GuVpc(stack, "MyVpc", { ssmParameters: true });
    expect(stack).toHaveResourceLike("AWS::SSM::Parameter", {
      Name: "/account/vpc/primary/id",
    });

    expect(stack).toHaveResourceLike("AWS::SSM::Parameter", {
      Name: "/account/vpc/primary/subnets/public",
    });

    expect(stack).toHaveResourceLike("AWS::SSM::Parameter", {
      Name: "/account/vpc/primary/subnets/private",
    });
  });

  it("should not create VPC SSM parameters if set to false", () => {
    const stack = simpleGuStackForTesting({ stack: "test-stack" });
    new GuVpc(stack, "MyVpc", { ssmParameters: false });

    expect(stack).toCountResources("AWS::SSM::Parameter", 0);
  });
});
