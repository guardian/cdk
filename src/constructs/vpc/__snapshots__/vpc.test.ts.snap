// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuVpc construct should match snapshot 1`] = `
Object {
  "Resources": Object {
    "MyVpcF9F0CA6F": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "MyVpcIGW5C4A4F63": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "MyVpcVPCGW488ACE0D": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "MyVpcapplicationSubnet1DefaultRoute4B900A14": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "MyVpcingressSubnet1NATGatewayC15F3CF1",
        },
        "RouteTableId": Object {
          "Ref": "MyVpcapplicationSubnet1RouteTableFCDD946D",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcapplicationSubnet1RouteTableAssociation37281D7C": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "MyVpcapplicationSubnet1RouteTableFCDD946D",
        },
        "SubnetId": Object {
          "Ref": "MyVpcapplicationSubnet1Subnet0A4FCBEC",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcapplicationSubnet1RouteTableFCDD946D": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet1",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcapplicationSubnet1Subnet0A4FCBEC": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.128.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "application",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet1",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcapplicationSubnet2DefaultRouteA21A20B8": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "MyVpcingressSubnet2NATGateway220B5C59",
        },
        "RouteTableId": Object {
          "Ref": "MyVpcapplicationSubnet2RouteTable1A5026C8",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcapplicationSubnet2RouteTable1A5026C8": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet2",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcapplicationSubnet2RouteTableAssociationD1C39589": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "MyVpcapplicationSubnet2RouteTable1A5026C8",
        },
        "SubnetId": Object {
          "Ref": "MyVpcapplicationSubnet2Subnet04E944F4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcapplicationSubnet2Subnet04E944F4": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.192.0/18",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "application",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/applicationSubnet2",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcdynamodb4933D0A6": Object {
      "Properties": Object {
        "RouteTableIds": Array [
          Object {
            "Ref": "MyVpcapplicationSubnet1RouteTableFCDD946D",
          },
          Object {
            "Ref": "MyVpcapplicationSubnet2RouteTable1A5026C8",
          },
          Object {
            "Ref": "MyVpcingressSubnet1RouteTableBD23564A",
          },
          Object {
            "Ref": "MyVpcingressSubnet2RouteTable17765BD0",
          },
        ],
        "ServiceName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "com.amazonaws.",
              Object {
                "Ref": "AWS::Region",
              },
              ".dynamodb",
            ],
          ],
        },
        "VpcEndpointType": "Gateway",
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "MyVpcingressSubnet1DefaultRoute5C0E02BD": Object {
      "DependsOn": Array [
        "MyVpcVPCGW488ACE0D",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "RouteTableId": Object {
          "Ref": "MyVpcingressSubnet1RouteTableBD23564A",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcingressSubnet1EIPD447827A": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet1",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "MyVpcingressSubnet1NATGatewayC15F3CF1": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "MyVpcingressSubnet1EIPD447827A",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "MyVpcingressSubnet1Subnet05B4D133",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet1",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "MyVpcingressSubnet1RouteTableAssociationCA312C68": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "MyVpcingressSubnet1RouteTableBD23564A",
        },
        "SubnetId": Object {
          "Ref": "MyVpcingressSubnet1Subnet05B4D133",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcingressSubnet1RouteTableBD23564A": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet1",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcingressSubnet1Subnet05B4D133": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet1",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcingressSubnet2DefaultRouteF6D3CDFA": Object {
      "DependsOn": Array [
        "MyVpcVPCGW488ACE0D",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "MyVpcIGW5C4A4F63",
        },
        "RouteTableId": Object {
          "Ref": "MyVpcingressSubnet2RouteTable17765BD0",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "MyVpcingressSubnet2EIP1342BA84": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet2",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "MyVpcingressSubnet2NATGateway220B5C59": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "MyVpcingressSubnet2EIP1342BA84",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "MyVpcingressSubnet2Subnet6E3FCA55",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet2",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "MyVpcingressSubnet2RouteTable17765BD0": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet2",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "MyVpcingressSubnet2RouteTableAssociation10D8624C": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "MyVpcingressSubnet2RouteTable17765BD0",
        },
        "SubnetId": Object {
          "Ref": "MyVpcingressSubnet2Subnet6E3FCA55",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "MyVpcingressSubnet2Subnet6E3FCA55": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.64.0/18",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "ingress",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          Object {
            "Key": "Name",
            "Value": "Test/MyVpc/ingressSubnet2",
          },
          Object {
            "Key": "Stack",
            "Value": "test-stack",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "MyVpcs3D9E23531": Object {
      "Properties": Object {
        "RouteTableIds": Array [
          Object {
            "Ref": "MyVpcapplicationSubnet1RouteTableFCDD946D",
          },
          Object {
            "Ref": "MyVpcapplicationSubnet2RouteTable1A5026C8",
          },
          Object {
            "Ref": "MyVpcingressSubnet1RouteTableBD23564A",
          },
          Object {
            "Ref": "MyVpcingressSubnet2RouteTable17765BD0",
          },
        ],
        "ServiceName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "com.amazonaws.",
              Object {
                "Ref": "AWS::Region",
              },
              ".s3",
            ],
          ],
        },
        "VpcEndpointType": "Gateway",
        "VpcId": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "privateSubnetsEA64D674": Object {
      "Properties": Object {
        "Name": "/account/vpc/primary/subnets/private",
        "Tags": Object {
          "Stack": "test-stack",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/cdk",
        },
        "Type": "StringList",
        "Value": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "MyVpcapplicationSubnet1Subnet0A4FCBEC",
              },
              ",",
              Object {
                "Ref": "MyVpcapplicationSubnet2Subnet04E944F4",
              },
            ],
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "publicSubnetsFE05F27B": Object {
      "Properties": Object {
        "Name": "/account/vpc/primary/subnets/public",
        "Tags": Object {
          "Stack": "test-stack",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/cdk",
        },
        "Type": "StringList",
        "Value": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "MyVpcingressSubnet1Subnet05B4D133",
              },
              ",",
              Object {
                "Ref": "MyVpcingressSubnet2Subnet6E3FCA55",
              },
            ],
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "vpcIDEC623FC3": Object {
      "Properties": Object {
        "Name": "/account/vpc/primary/id",
        "Tags": Object {
          "Stack": "test-stack",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/cdk",
        },
        "Type": "String",
        "Value": Object {
          "Ref": "MyVpcF9F0CA6F",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;
